aziotd/src/main.rs

```diff
@@ -186,58 +227,154 @@ where
 }
 
 fn merge_toml(base: &mut toml::Value, patch: toml::Value) {
-    // Similar to JSON patch, except that maps are called tables, and
-    // there is no equivalent of null that can be used to remove keys from an object.
+    // Similar to JSON patch, except that:
+    //
+    // - Maps are called tables.
+    // - There is no equivalent of null that can be used to remove keys from an object.
+    // - Arrays are merged via concatenating the patch to the base, rather than replacing the base with the patch.
 
-    if let toml::Value::Table(base) = base {
+    if let toml::Value::Table(base) = &mut *base {
         if let toml::Value::Table(patch) = patch {
-            for (k, v) in patch {
-                merge_toml(base.entry(k).or_insert(toml::Value::Boolean(false)), v);
+            for (key, value) in patch {
+                // Insert a dummy `false` if the original key didn't exist at all. It'll be overwritten by `value` in that case.
+                let original_value = base.entry(key).or_insert(toml::Value::Boolean(false));
+                merge_toml(original_value, value);
             }
 
             return;
         }
     }
 
-    *base = patch;
-}
-
-#[derive(Debug)]
-struct Error(ErrorKind, backtrace::Backtrace);
+    if let toml::Value::Array(base) = base {
+        if let toml::Value::Array(patch) = patch {
+            base.extend(patch);
+            return;
+        }
+    }
 
-#[derive(Debug)]
-enum ErrorKind {
-    GetProcessName(std::borrow::Cow<'static, str>),
-    ReadConfig(Option<std::path::PathBuf>, Box<dyn std::error::Error>),
-    Service(Box<dyn std::error::Error>),
+    *base = patch;
 }
 
-impl std::fmt::Display for ErrorKind {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        match self {
-            ErrorKind::GetProcessName(message) => write!(f, "could not read argv[0]: {}", message),
-            ErrorKind::ReadConfig(Some(path), _) => {
-                write!(f, "could not read config from {}", path.display())
-            }
-            ErrorKind::ReadConfig(None, _) => f.write_str("could not read config"),
-            ErrorKind::Service(_) => f.write_str("service encountered an error"),
+#[cfg(test)]
+mod tests {
+    #[test]
+    fn process_name_from_args() {
+        for &(input, expected) in &[
```

Comment:
super minor, but it would be nice to avoid repeating `#[cfg(debug_assertions)]` by using a Vec. i.e:

```rust
let mut test_cases = Vec::new();
test_cases.extend(&[ /* ... */ ]);
#[cfg(debug_assertions)]
test_cases.extend(&[ /* ... */ ])
```

Reply:
Yeah, I was just avoiding an allocation, but it's not a big deal for a test.

Done.